src/CommandLineUtils/Abstractions/DefaultValueParserFactory.cs

```diff
@@ -0,0 +1,123 @@
+﻿namespace McMaster.Extensions.CommandLineUtils.Abstractions
+{
+    using System;
+    using System.Collections.Generic;
+    using System.ComponentModel;
+    using System.Diagnostics;
+    using System.Globalization;
+    using System.Text;
+
+    /// <summary>
+    /// A factory creating generic implementations of <see cref="IValueParser{T}"/>. The implementations are based
+    /// on automatically located <see cref="TypeConverter"/> classes that are suitable for parsing.
+    /// </summary>
+    sealed class DefaultValueParserFactory
+    {
+        const int DefaultMaxCacheCapacity = 100;
+
+        #region Private Fields
+
+        private readonly int _maxCacheCapacity;
+
+        [DebuggerBrowsable(DebuggerBrowsableState.RootHidden)]
+        readonly Dictionary<Type, IValueParser> _parsersByTargetType = new Dictionary<Type, IValueParser>();
+
+        [DebuggerBrowsable(DebuggerBrowsableState.Never)]
+        readonly HashSet<Type> _notSupportedTypes = new HashSet<Type>();
+
+        #endregion
+
+        public DefaultValueParserFactory(int maxCacheCapacity = DefaultMaxCacheCapacity)
+        {
+            _maxCacheCapacity = maxCacheCapacity;
+            if (_maxCacheCapacity < 1) throw new ArgumentOutOfRangeException(nameof(maxCacheCapacity));
+        }
+
+        [DebuggerStepThrough]
+        public bool TryGetParser<T>(out IValueParser<T> parser)
+        {
+            if (TryGetParser<T>(out IValueParser generalizedParser))
+            {
+                parser = (IValueParser<T>)generalizedParser;
+                return true;
+            }
+
+            parser = null;
+            return false;
+        }
+
+        public bool TryGetParser<T>(out IValueParser parser)
+        {
+            var targetType = typeof(T);
+            if (_notSupportedTypes.Contains(targetType))
+            {
+                parser = null;
+                return false;
+            }
+
+            if (_parsersByTargetType.TryGetValue(targetType, out parser))
+            {
+                Debug.Assert(targetType == parser.TargetType);
+                return true;
+            }
+
+            var converter = TypeDescriptor.GetConverter(targetType);
+            if (converter.CanConvertFrom(typeof(string)))
+            {
+                if (_parsersByTargetType.Count >= _maxCacheCapacity)
```

Comment:
Why is there a limit to 100 type converters?

Reply:
Limiting the cache capacity prevents the excessive memory consumption under stress loads or invocations triggered by a malfunctioning caller. I subjectively chosen 100 as the requirement to support more than 100 different argument types per command line application is unlikely. Still, the component is capable of handling these exotic cases with some performance penalty (approximately 3 times slower).

Reply:
Spoken like a perf-minded engineer. :) I see where you are coming from, but I think this might be premature optimization. It appears that the implementation of TypeDescriptor is already caching its results and has some memory optimizations under the hood (like using WeakReference). (see https://github.com/dotnet/runtime/blob/master/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs) I think we can simplify this code by skipping the cache and count checks.

Reply:
It is fine by me. This component is private anyway.